torchrl/envs/batched_envs.py

```diff
@@ -832,21 +835,19 @@ def _step(self, tensordict: TensorDictBase) -> TensorDictBase:
 
     @_check_start
     def _reset(self, tensordict: TensorDictBase, **kwargs) -> TensorDictBase:
-
-        missing_reset = False
         if tensordict is not None:
-            needs_resetting = [False for _ in range(self.num_workers)]
-            for _reset_key in self.reset_keys:
-                _reset = tensordict.get(_reset_key, default=None)
-                if _reset is not None:
-                    for i in range(self.num_workers):
-                        needs_resetting[i] += _reset[i].any()
-                else:
-                    missing_reset = True
-                    break
+            needs_resetting = _bring_reset_to_root(tensordict)
+            if needs_resetting.ndim > 2:
+                needs_resetting = needs_resetting.flatten(1, needs_resetting.ndim - 1)
+            if needs_resetting.ndim > 1:
+                needs_resetting = needs_resetting.any(-1)
+            elif not needs_resetting.ndim:
+                needs_resetting = needs_resetting.expand(self.batch_size)
+        else:
+            needs_resetting = torch.ones(
+                (self.num_workers,), device=self.device, dtype=torch.bool
+            )
```

Comment:
?

Reply:
why do we switch on the ndim, couldn't we simply squash it into 1 dim?
also expanding to self.batch_size seems wrong to me as that could be bigger than self.n_workers.

Reply:
Sorry I'm still not following at all what the first comment is about, so let me explain what we're doing here:
If there is an input data structure, we gather the reset info.

We pass the reset_keys and obtain a reset of the size of the tensordict. This is the contract of _aggregate_resets.
The goal is to support arrays of envs that are more complex than just a list of envs.
In our case, if we have more than one dimension, we flatten and call any on the last dimension to obtain something that has the shape [n_workers]. And in case we don't have any element (which can happen) we expand to the appropriate shape. You can check that this last condition is necessary by running the tests but essentially the idea is just that tensordict can miss the `"_reset"` key so the funciton just returns a scalar.
Now sure we can expand to the num_workers that's safer!

Reply:
my point is just that if `needs_resetting` has to have shape [n_workers]
`needs_resetting = needs_resetting.expand(self.batch_size)` will not do that (because the batch size can be something else)